pkg/uefi/meregion.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2018 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uefi
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"log"
+)
+
+// ME Partition parsing, the goal is to spot a padding in the ME Region
+// after the ME partitions so that this region can be shrunk in the IFD.
+//
+// ME partition informations from http://me.bios.io/ME_blob_format
+
+// MEFTPSignature is the sequence of bytes that an ME Flash Partition
+// Table is expected to start with.
+var (
+	MEFTPSignature = []byte{0x24, 0x46, 0x50, 0x54}
+)
+
+const (
+	// MEFTPSignatureLength represents the size of the ME FTP signature
+	MEFTPSignatureLength = 4
+	// MEPartitionTableEntryLength is the size of a partition table entry
+	MEPartitionTableEntryLength = 32
+)
+
+// MEFPT is the main structure that represents an ME Flash Partition Table.
+type MEFPT struct {
+	// Holds the raw buffer
+	buf []byte
+
+	PartitionCount    uint32
+	PartitionMapStart int
+	Entries           []MePartitionEntry
+	//Metadata for extraction and recovery
+	ExtractPath string
+}
+
+// MePartitionEntry is an entry in FTP
+type MePartitionEntry struct {
+	Name     MeName
+	Owner    [4]byte
+	Offset   uint32
+	Length   uint32
+	Reserved [4]uint32
+}
+
+// MeName represent 4 bytes with JSON string support
+// OK this is probably overkill!
+type MeName [4]byte
+
+// MarshalText converts MeName to a byte range (for JSON)
+func (n MeName) MarshalText() ([]byte, error) {
+	e := len(n)
+	for e > 0 && n[e-1] == 0 {
+		e--
+	}
+	return n[:e], nil
+}
+
+// UnmarshalText converts a byte range to MeName (for JSON)
+func (n *MeName) UnmarshalText(b []byte) error {
+	for i := range n[:] {
```

Comment:
It should be possible to use Go's builtin `copy` function to copy from `b` to `n`. Something like `copy((*n)[:], b)`. The copy function uses the minimum size of both slices, so this should be safe whatever the size are.

Reply:
I see but the copy will not fill the remaining bytes with 0 when b is shorter than 4.
The Zero value should be ok, but can UnmarshalText be called on a reused MEName ?
... well I'll add some test and check that.

Reply:
I endedup creating a new MEName, using `copy` and then affecting it to the receiver.
It might be too much care, but MEName is only 4 bytes anyway so it does not cost much !